programs/token/src/instructions/transfer_checked.rs

```diff
@@ -0,0 +1,76 @@
+use core::mem::MaybeUninit;
+
+use pinocchio::{
+    account_info::AccountInfo,
+    instruction::{AccountMeta, Instruction, Signer},
+    program::invoke_signed,
+    ProgramResult,
+};
+
+/// Transfer Tokens from one Token Account to another.
+///
+/// ### Accounts:
+///   0. `[WRITE]` The source account.
+///   1. `[]` The token mint.
+///   2. `[WRITE]` The destination account.
+///   3. `[SIGNER]` The source account's owner/delegate.
+pub struct TransferChecked<'a> {
+    /// Sender account.
+    pub from: &'a AccountInfo,
+    /// Mint Account
+    pub mint: &'a AccountInfo,
+    /// Recipient account.
+    pub to: &'a AccountInfo,
+    /// Authority account.
+    pub authority: &'a AccountInfo,
+    /// Amount of microtokens to transfer.
+    pub amount: u64,
+    /// Decimal for the Token
+    pub decimals: u8,
+}
+
+impl<'a> TransferChecked<'a> {
+    #[inline(always)]
+    pub fn invoke(&self) -> ProgramResult {
+        self.invoke_signed(&[])
+    }
+
+    pub fn invoke_signed(&self, signers: &[Signer]) -> ProgramResult {
+        // account metadata
+        let account_metas: [AccountMeta; 4] = [
+            AccountMeta::writable(self.from.key()),
+            AccountMeta::readonly(self.mint.key()),
+            AccountMeta::writable(self.to.key()),
+            AccountMeta::readonly_signer(self.authority.key()),
+        ];
+
+        // Instruction data layout:
+        // -  [0]: instruction discriminator 
+        // -  [1..9]: amount 
+        // -  [9]: decimals
+        let mut instruction_data = MaybeUninit::<[u8; 10]>::uninit();
+
+        // Populate data
+        unsafe {
+            let ptr = instruction_data.as_mut_ptr() as *mut u8;
+            // Set discriminator as u8 at offset [0]
+            *ptr = 12;
```

Comment:
Unfortunately this is UB. The reason is that when you assign a value to `*ptr`, it will call `drop` on the previous value so creating a reference to uninitialized memory is UB:

https://doc.rust-lang.org/beta/std/mem/union.MaybeUninit.html#method.as_mut_ptr

I think it would be fine to do the static allocation of the array:
```
let mut instruction_data = [0u8; 10];
```